projects/com.oracle.truffle.llvm.parser/src/com/oracle/truffle/llvm/parser/util/LLVMControlFlowGraph.java

```diff
@@ -0,0 +1,335 @@
+            for (CFGBlock b : successors) {
+                sucIDs[i++] = b.id;
+            }
+
+            return sucIDs;
+        }
+
+        public boolean isInLoop(CFGBlock block) {
+            return block == loopHeader || body.contains(block);
+        }
+
+        public CFGBlock getHeader() {
+            return loopHeader;
+        }
+
+        @Override
+        public String toString() {
+            return "Loop: " + this.id + " - Header: " + this.loopHeader.toString();
+        }
+
+    }
+
+    public LLVMControlFlowGraph(InstructionBlock[] blocks) {
+        this.blocks = new CFGBlock[blocks.length];
+        for (int i = 0; i < blocks.length; i++) {
+            this.blocks[i] = new CFGBlock(blocks[i]);
+        }
+    }
+
+    private void resolveEdges() {
+        for (CFGBlock block : blocks) {
+            TerminatingInstruction term = block.instructionBlock.getTerminatingInstruction();
+            // set successors and predecessors
+            for (int i = 0; i < term.getSuccessorCount(); i++) {
+                int sucId = term.getSuccessor(i).getBlockIndex();
+                block.sucs.add(this.blocks[sucId]);
+                this.blocks[sucId].preds.add(block);
+            }
+        }
+
+    }
+
+    public List<CFGLoop> getCFGLoops() {
+        List<CFGLoop> loops = new ArrayList<>();
+
+        for (CFGLoop l : cfgLoops) {
+            if (l != null)
+                loops.add(l);
+        }
+        loops.removeIf(l -> l == null);
+        return loops;
+    }
+
+    public boolean isReducible() {
+        return reducible;
+    }
+
+    public void build() {
+        // set successors and predecessors
+        resolveEdges();
+
+        long openLoops = openLoops(blocks[0]);
+        if (openLoops != 0) {
+            reducible = false;
+            throw new RuntimeException("Irreducible control flow!"); // bailout
+        }
+
+        sortLoops();
+
+        for (CFGLoop l : getCFGLoops()) {
+            l.calculateSuccessors();
+        }
+    }
+
+    private boolean sortLoops() {
+        List<CFGLoop> sorted = new ArrayList<>();
+        List<CFGLoop> active = new ArrayList<>();
+
+        for (CFGLoop l : getCFGLoops()) {
+            sortLoop(sorted, active, l);
+        }
+
+        cfgLoops = sorted.toArray(new CFGLoop[cfgLoops.length]);
+        return true;
+    }
+
+    private void sortLoop(List<CFGLoop> sorted, List<CFGLoop> active, CFGLoop loop) {
+        if (sorted.contains(loop))
+            return;
+
+        active.add(loop);
+        for (CFGBlock b : loop.body) {
+            if (b.isLoopHeader) {
+                CFGLoop inner = cfgLoops[b.loopId];
+                if (active.contains(inner)) {
+                    // catches case that there is a stack overflow because two loop nodes are being called iteratively
+                    // from one another, without one being left beforehand
+                    throw new RuntimeException("Irreducible nestedness!");
```

Comment:
throw a custom exception and catch it in the build method (applies to all bailouts in here)